lib/common/alerts.c

```diff
@@ -0,0 +1,191 @@
+#include <crm/msg_xml.h>
+#include <crm/common/alerts_internal.h>
+
+
+char *notify_script = NULL;
+char *notify_target = NULL;
+GListPtr notify_list = NULL;
+guint max_alert_timeout = CRMD_NOTIFY_DEFAULT_TIMEOUT_MS;
+
+/*		
+ * to allow script compatibility we can have more than one		
+ * set of environment variables		
+ */
+const char *notify_keys[14][3] =		
+{		
+    [CRM_notify_recipient]     = {"CRM_notify_recipient",     "CRM_alert_recipient",     NULL},		
+    [CRM_notify_node]          = {"CRM_notify_node",          "CRM_alert_node",          NULL},		
+    [CRM_notify_nodeid]        = {"CRM_notify_nodeid",        "CRM_alert_nodeid",        NULL},		
+    [CRM_notify_rsc]           = {"CRM_notify_rsc",           "CRM_alert_rsc",           NULL},		
+    [CRM_notify_task]          = {"CRM_notify_task",          "CRM_alert_task",          NULL},		
+    [CRM_notify_interval]      = {"CRM_notify_interval",      "CRM_alert_interval",      NULL},		
+    [CRM_notify_desc]          = {"CRM_notify_desc",          "CRM_alert_desc",          NULL},		
+    [CRM_notify_status]        = {"CRM_notify_status",        "CRM_alert_status",        NULL},		
+    [CRM_notify_target_rc]     = {"CRM_notify_target_rc",     "CRM_alert_target_rc",     NULL},		
+    [CRM_notify_rc]            = {"CRM_notify_rc",            "CRM_alert_rc",            NULL},		
+    [CRM_notify_kind]          = {"CRM_notify_kind",          "CRM_alert_kind",          NULL},		
+    [CRM_notify_version]       = {"CRM_notify_version",       "CRM_alert_version",       NULL},		
+    [CRM_notify_node_sequence] = {"CRM_notify_node_sequence", "CRM_alert_node_sequence", NULL},		
+    [CRM_notify_timestamp]     = {"CRM_notify_timestamp",     "CRM_alert_timestamp",     NULL}		
+};
+
+static void		
+free_envvar_entry(envvar_t *entry)		
+{		
+    free(entry->name);		
+    free(entry->value);		
+    free(entry);		
+}		
+
+static void		
+free_notify_list_entry(notify_entry_t *entry)		
+{		
+    free(entry->id);		
+    free(entry->path);		
+    free(entry->tstamp_format);		
+    free(entry->recipient);		
+    if (entry->envvars) {		
+        g_list_free_full(entry->envvars,		
+                         (GDestroyNotify) free_envvar_entry);		
+    }		
+    free(entry);		
+}		
+
+void		
+free_notify_list()		
+{		
+    if (notify_list) {		
+        g_list_free_full(notify_list, (GDestroyNotify) free_notify_list_entry);		
+        notify_list = NULL;		
+    }		
+}		
+
+static gpointer		
+copy_envvar_entry(envvar_t * src,		
+                  gpointer data)		
+{		
+    envvar_t *dst = calloc(1, sizeof(envvar_t));		
+
+    CRM_ASSERT(dst);		
+    dst->name = strdup(src->name);		
+    dst->value = src->value?strdup(src->value):NULL;		
+    return (gpointer) dst;		
+}		
+
+GListPtr		
+add_dup_envvar(GListPtr envvar_list,		
+               envvar_t *entry)		
+{		
+    return g_list_prepend(envvar_list, copy_envvar_entry(entry, NULL));		
+}		
+
+GListPtr		
+drop_envvars(GListPtr envvar_list, int count)		
```

Comment:
It might be convenient to make these functions take envar_t *entry as an argument instead of envvar_list. The function can use entry->envvars.